### PR TITLE
#162163005 Enable user to comment on an article.

### PR DESCRIPTION
**Pull request by kizzanaome**
### What does this PR do?
- Allows a user to comment on an article.
- Enables a user to reply to a comment.

### Description of Task to be completed?
- A  user should be able to add a comment(s) on an article and he should also  be able to reply on that particular comment.

### How should this be manually tested?
Have this endpoints working.
- Post a single comment.
``` POST:  /api/<slug>/comments```

- get  comments of an article
```GET: /api/<slug>/comments```

- Get a single comment
``` GET: /api/<slug>/comments/<id>```
- delete a comment
```DELETE : /api/:slug/comments/:id```
- Reply to a comment
``` GET: /api/<slug>/comments/<id>/thread```

First clone this repository
```
$ git clone https://github.com/andela/ah-backend-poseidon
$ cd ah-backend-poseidon
```
Create virtual environment and install it
```
$ python3 -m venv venv
$ source/venv/bin/activate
```
Then install all the necessary dependencies
```
pip install -r requirements.txt
```
Run the application
At the terminal or console type
```
 $ python manage.py make migrations
 $ python manage.py migrate
 $ python manage.py run server
```

Head over to postman and run a Get method to view an article through this endpoint /api/articles/. 
You will be able to view an article like this one below. 

```{
    "articles": {
        "slug": "who-is-he-he-has-20181212152330656837",
        "title": "Who is he?",
        "description": "He has a bald head",
        "body": "It is believed he will rule Uganda to heaven",
        "created_on": "2018-12-12T15:23:30.657217Z",
        "average_rating": 0,
        "updated_on": "2018-12-12T15:23:30.657248Z",
        "image_url": null,
        "author": {
            "username": "naomecoG",
            "email": "naomekizzan@gmail.com",
            "bio": "",
            "image": null,
            "following": false
        },
        "tags": [
            "programming",
            "technology",
            "python"
        ]
    }
}
```

On this article, copy the slug that has been returned and paste it on thhe comment url like this /api/who-is-he-he-has-20181212152330656837/comments .Use this endpoint to sent a post request with the endpoint. 
Use the formart bellow to type a comment of your choice .

```
{
	"comment":{
		"body":"This is a comment"
	}
}
```
Send a get method to view the comment that you posted through this endpoint.

### What are the relevant pivotal tracker stories?
 #30

### CODE REVIEW

**Review by arthurarty — CHANGES_REQUESTED** [review]
PR title is not following the standard. Please check previous prs. Also this pr has dropped our coverage. Please add more tests.

**Review by okellogabrielinnocent — COMMENTED** [review]
- Change title to meet the organization conventions
- Write more tests to raise your tests as they have gone below standards
- Add endpoint for updating comment
- Do not upload migrations to the code base
- Add description to commands on how to manually test it.
- Add comments to the code i.e on serializers.py, views.py and test files
- Always remember to remove your password on code committed

**Review by okellogabrielinnocent — COMMENTED** [review]
Good work done

File: authors/apps/comments/renderer.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        # the default JSONRenderer to handle rendering errors, so we need to
+        # check for this case.
+
+        if type(data) != ReturnList:  # try returning dict
+            errors = data.get('errors', None)
+            if errors is not None:
+                # As mentioned about, we will let the default JSONRenderer handle
+                # rendering errors.
+                return super(CommentJSONRenderer, self).render(data)
+        # Finally, we can render our data under the "user" namespace.
```
Comment (Luleherll) [inline authors/apps/comments/renderer.py:21]: Change the comment to what the renderer actually returns.

File: authors/apps/tests/test_articles.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def test_get_all_articles(self):
         self.assertEqual(len(response.json().get("articles")), 3)
 
     def test_get_specific_article(self):
-        slug = self.slugger()
+        slug = self.slugger2()
```
Comment (malep2007) [inline authors/apps/tests/test_articles.py:50]: If slugger was deleted, why do we do have slugger2()?

File: authors/apps/comments/serializers.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        model = Comment
+        fields = ('id', 'body',  'commented_by',
+                  'created_at', 'updated_at', 'replies', 'parent')
+
+    def get_is_parent(self, object):
+        if not object.is_parent:
+            return CommentChildSerializer(object.children(), many=True).data
+        return object.parent
+
+    def validate(self, data):
```
Comment (malep2007) [inline authors/apps/comments/serializers.py]: Refactor the `validate` method, create a helper method to reduce duplication. Line 24 and line 41

File: authors/apps/comments/urls.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from django.urls import path
+from .views import (
+    CommentCreateListView,CommentsListThreadsCreateView, CommentsAPIView)
+
+
+app_name = 'comments'
+urlpatterns = [
+    path('<slug>/comments/', CommentCreateListView.as_view(), name='comments'),
+    path('<slug>/comments/<int:id>', CommentsAPIView.as_view(), name='single_comment'),
+    path('<slug>/comments/<int:id>/thread', CommentsListThreadsCreateView.as_view(),name='thread comment'),
```
Comment (arthurarty) [inline authors/apps/comments/urls.py]: On line 10 name ='thread comment' while when I look at line 9. name='single_comment'. So line 10 would have been something like name='thread_comment' and therefore consistency.

**Review by malep2007 — CHANGES_REQUESTED** [review]
There should be clear communication if you are modifying work that was previously done by someone else. This is a context-specific example of stakeholder management. I was not informed of possible changes and reasons for the implementation.

**Comment by malep2007** [comment]
Test Coverage has reduced. Please write more tests as indicated by @okellogabrielinnocent .

**Comment by aayebare** [comment]
could we add doc_strings to some of these functions that don't have?
Otherwise,great work

File: authors/apps/comments/urls.py (hunk 4 above)
Comment (kizzanaome) [reply authors/apps/comments/urls.py]: I have worked on the feedback @arthurarty

File: authors/apps/comments/serializers.py (hunk 3 above)
Comment (kizzanaome) [reply authors/apps/comments/serializers.py]: @malep2007 I have refactored the validated method as requested

File: authors/apps/tests/test_articles.py (hunk 2 above)
Comment (kizzanaome) [reply authors/apps/tests/test_articles.py:50]: @malep2007  I apologise for the communication I said earlier, The slugger function was not deleted, we just created another slugger helper method which has a verified user.unlkie the first one.

**Review by arthurarty — APPROVED** [review]

**Review by okellogabrielinnocent — APPROVED** [review]

**Review by billkabanga — APPROVED** [review]

**Review by aayebare — APPROVED** [review]

**Review by malep2007 — APPROVED** [review]
ok

**Pull request merged**